### DISCUSSION TITLE
Migration generator keeps altering CHAR columns in Postgresql

### DISCUSSION THREAD

**Discussion by arminrosu** [mikro-orm/mikro-orm#4696] [Q&A]
Hey,

I'm having an issue with the migration generator which I can't figure out. 

**Schema**

```ts
export const AppleSchema = new EntitySchema({
  class: Apple,
  properties: {
    id: {
      primary: true,
      length: 50,
      // `char` has same effect, but the column doesn't get the length set:
      // type: 'char',
      type: 'char(50)',
    }
  },
  repository: () => AppleRepository,
  schema: 'graph$default',
  tableName: 'Apple',
})
```

**Previous Migration**

```sql
CREATE TABLE
  "graph$default"."Apple" (
    "id" CHAR(50) NOT NULL,
    CONSTRAINT "Apple_pkey" PRIMARY KEY ("id")
  );
```

Every subsequent call of `npx mikro-orm migration:create` will generate this migration SQL:

```sql
ALTER TABLE "graph$default"."Apple"
ALTER COLUMN "id"
TYPE CHAR(50) USING ("id"::CHAR(50));
```

This happens no matter how often I apply `migration:up` or call `migration:create` again. It doesn't happen for non-string type fields.

Software versions:
- mikro-orm: 5.8
- postgresql: 13.0, 15.4 (via docker, from the official images)
- node: 16.20.1

Would appreciate any help, I'm not sure what's wrong.

**Comment by B4nan** [comment]
You can enable extensive logging for the CLI, which will include why the schema comparator thinks the property needs changes.

```bash
MIKRO_ORM_VERBOSE=1 npx mikro-orm migration:create
```

Also, it's probably better to test this with `schema:update -d` instead of going through migrations, as those will work based on the schema snapshot (basically a cache for the last known database state).

**Comment by arminrosu** [comment]
Ah, thanks for the swift response. Here's the output:

```
'type' changed for column graph$default.Apple.id { columnType1: 'bpchar', columnType2: 'char(50)' }
```

Sounds a bit like https://github.com/mikro-orm/mikro-orm/discussions/2791 - they are supposed to be aliases which mikro-orm doesn't recognize. There's a suggested workaround in the discussion above, I'll try that first.

Thanks again for the quick help!

**Comment by arminrosu** [comment]
`MIKRO_ORM_VERBOSE=1` is the bees knees!

```
[schema] 'type' changed for column graph$default.Apple.id { columnType1: 'bpchar', columnType2: 'bpchar(50)' }
```

So changing the schema to:

```ts
  properties: {
    id: {
      primary: true,
      length: 50,
      type: 'bpchar',
    }
  },
```

Leads to:

```
No changes required, schema is up-to-date
```

🥳 

But now the columns have no fixed length, ie. the length property is not used to generate the migration. I'll dig deeper, see what I can do.

**Reply by B4nan** [reply]
Try this:

```ts
  properties: {
    id: {
      primary: true,
      columnType: 'bpchar(50)',
    }
  },
```

**Reply by arminrosu** [reply]
Nope, same behaviour as before - it will constantly create new migrations, as the length is part of the type: 

```
'type' changed for column graph$default.Account_ApiKey.clientId { columnType1: 'bpchar', columnType2: 'bpchar(50)' }
```

---

Sorry, just noticed I was using `type` instead of `columnType` in the samples above. I actually was always using `columnType`.

**Reply by arminrosu** [reply]
I change the column type from `bpchar` to `varchar` and added the length in the `columnType` property. It now works. Very strange.

```ts
columnType: 'varchar(50)',
length: 50
```

The `length` property seems to have no effect on the actual column creation / altering.

**Reply by B4nan** [reply]
> The length property seems to have no effect on the actual column creation / altering.

Not if you use `columnType` as that is supposed to be complete, including the length. The `length` option is valid only if you don't use `columnType`.
